File: api/application.go
package api

import (
	"time"

	"github.com/distr-sh/distr/internal/types"
	"github.com/google/uuid"
)

type ApplicationResponse struct {
	ID                    uuid.UUID                    `json:"id"`
	CreatedAt             time.Time                    `json:"createdAt"`
	Name                  string                       `json:"name"`
	Type                  types.DeploymentType         `json:"type"`
	ImageID               *uuid.UUID                   `json:"imageId,omitempty"`
	ImageUrl              *string                      `json:"imageUrl,omitempty"`
	VersioningStrategy    types.VersioningStrategy     `json:"versioningStrategy"`
	AllowAutomaticUpdates bool                         `json:"allowAutomaticUpdates"`
	Versions              []ApplicationVersionResponse `json:"versions"`
}

type ApplicationVersionResponse struct {
	ID            uuid.UUID                  `json:"id"`
	CreatedAt     time.Time                  `json:"createdAt"`
	ArchivedAt    *time.Time                 `json:"archivedAt,omitempty"`
	Name          string                     `json:"name"`
	LinkTemplate  string                     `json:"linkTemplate"`
	ApplicationID uuid.UUID                  `json:"applicationId"`
	ChartType     *types.HelmChartType       `json:"chartType,omitempty"`
	ChartName     *string                    `json:"chartName,omitempty"`
	ChartUrl      *string                    `json:"chartUrl,omitempty"`
	ChartVersion  *string                    `json:"chartVersion,omitempty"`
	CreatedBy     *ApplicationVersionCreator `json:"createdBy,omitempty"`
}

type ApplicationVersionCreator struct {
	ID      uuid.UUID  `json:"id"`
	Name    string     `json:"name,omitempty"`
	Email   string     `json:"email,omitempty"`
	ImageID *uuid.UUID `json:"imageId,omitempty"`
}

type CreateApplicationRequest struct {
	Name string               `json:"name"`
	Type types.DeploymentType `json:"type"`
	// VersioningStrategy defaults to chronological when absent, which imposes no requirement on
	// version names and so cannot break a client written before the field existed.
	VersioningStrategy    types.VersioningStrategy `json:"versioningStrategy,omitempty"`
	AllowAutomaticUpdates bool                     `json:"allowAutomaticUpdates,omitempty"`
}

type UpdateApplicationRequest struct {
	Name               string                   `json:"name"`
	VersioningStrategy types.VersioningStrategy `json:"versioningStrategy,omitempty"`
	// AllowAutomaticUpdates leaves the application's setting alone when it is absent, so that a
	// client written before the field existed cannot turn automatic updates off.
	AllowAutomaticUpdates *bool `json:"allowAutomaticUpdates,omitempty"`
}

type PatchApplicationRequest struct {
	Name                  *string                          `json:"name,omitempty"`
	VersioningStrategy    *types.VersioningStrategy        `json:"versioningStrategy,omitempty"`
	AllowAutomaticUpdates *bool                            `json:"allowAutomaticUpdates,omitempty"`
	Versions              []PatchApplicationVersionRequest `json:"versions,omitempty"`
}

type PatchApplicationVersionRequest struct {
	ID         uuid.UUID  `json:"id"`
	ArchivedAt *time.Time `json:"archivedAt,omitempty"`
}

type CreateApplicationVersionRequest struct {
	Name         string                              `json:"name"`
	LinkTemplate string                              `json:"linkTemplate,omitempty"`
	ChartType    *types.HelmChartType                `json:"chartType,omitempty"`
	ChartName    *string                             `json:"chartName,omitempty"`
	ChartUrl     *string                             `json:"chartUrl,omitempty"`
	ChartVersion *string                             `json:"chartVersion,omitempty"`
	Resources    []ApplicationVersionResourceRequest `json:"resources,omitempty"`
}

type ApplicationVersionResourceRequest struct {
	Name string `json:"name"`
	// The content is rendered as markdown, where leading whitespace is syntax.
	Content            string `json:"content" trim:"-"`
	VisibleToCustomers bool   `json:"visibleToCustomers"`
}

type UpdateApplicationVersionRequest struct {
	Name       string     `json:"name"`
	ArchivedAt *time.Time `json:"archivedAt,omitempty"`
}

File: sdk/js/src/types/application.ts
import {BaseModel, Named} from './base';
import {DeploymentType, HelmChartType} from './deployment';

/**
 * How the latest version of an application is determined.
 * * 'semver' orders versions by semantic versioning, which every version name must follow.
 * * 'chronological' orders versions by their creation date.
 * * 'legacy' orders by semantic versioning where every name follows it and by creation date
 *   otherwise. It applies to applications created before a strategy existed and cannot be set.
 */
export type VersioningStrategy = 'semver' | 'chronological' | 'legacy';

export interface Application extends BaseModel, Named {
  type: DeploymentType;
  imageId?: string;
  imageUrl?: string;
  versioningStrategy?: VersioningStrategy;
  /** Whether deployments of this application may have automatic updates enabled. */
  allowAutomaticUpdates?: boolean;
  versions?: ApplicationVersion[];
}

export interface ApplicationVersion {
  id?: string;
  name: string;
  linkTemplate?: string;
  createdAt?: string;
  archivedAt?: string;
  applicationId?: string;
  chartType?: HelmChartType;
  chartName?: string;
  chartUrl?: string;
  chartVersion?: string;
  /** The user who created this version. Only vendor users get to see it. */
  createdBy?: ApplicationVersionCreator;
  resources?: ApplicationVersionResource[];
}

export interface ApplicationVersionCreator {
  id: string;
  name?: string;
  email?: string;
  imageId?: string;
}

export interface ApplicationVersionResource {
  id?: string;
  applicationVersionId?: string;
  name: string;
  content: string;
  visibleToCustomers: boolean;
}

export interface PatchApplicationRequest {
  name?: string;
  versioningStrategy?: VersioningStrategy;
  allowAutomaticUpdates?: boolean;
  versions?: {id: string; archivedAt?: string}[];
}
